test_python/test_tc.py

```diff
@@ -12,89 +12,499 @@
+        group_normalization = """
+            def moments(float(N, G, D, H, W) I) -> (mean, var) {
+                # var = E(x^2) - mean^2.
+                mean(n, g) +=! I(n, g, r_d, r_h, r_w)
+                 var(n, g) +=! I(n, g, r_d, r_h, r_w) * I(n, g, r_d, r_h, r_w)
+                mean(n, g)  = mean(n, g) / (D * H * W)
+                 var(n, g)  =  var(n, g) / (D * H * W) - mean(n, g) * mean(n, g)
+            }
+
+            def group_normalization(
+                float(N, G, D, H, W) I, float(G, D) gamma, float(G, D) beta,
+                float(N, G) mean, float(N, G) var) -> (O)
+            {
+                O(n, g, d, h, w) = gamma(g, d)
+                    * ( I(n, g, d, h, w) - mean(n, g) )
+                    * rsqrt( var(n, g) + 1e-5 )
+                    + beta(g, d)
+            }
+        """
+
+        N, G, D, H, W = 32, 32, 4, 56, 56
+        cache_filename = "/tmp/" + str(uuid.uuid4())
+        group_norm = tc.define_with_autograd(
+            group_normalization,
+            forward_entry_points=("moments", "group_normalization"),
+            backward_entry_points=(),
+            # First: I, gamma, beta -> I
+            #        => (0, )
+            # Second: I, gamma, beta, mean, var -> I, gamma, beta, mean, var
+            #        => (0, 1, 2, 3, 4) => None
+            forward_input_indices=((0, ), ),
+            cache_filename=cache_filename,
+            tuner_config=tuner_config)
+        I = Variable(torch.randn(N, G, D, H, W).cuda())
+        gamma = Variable(torch.randn(G, D).cuda())
+        beta = Variable(torch.randn(G, D).cuda())
+        # First occurrence on a new cache filename triggers tuning
+        mean, var, out = group_norm(I, gamma, beta)
+        # Subsequent occurrences do not
+        mean, var, out = group_norm(I, gamma, beta)
+        # TODO: test results!!!
+
+    #
+    # This test implements single kernel forward and backward with
+    # tc.define_with_autograd. This is the "light touch" mode and the function call
+    # may already be too verbose.
+    # Maybe consider making cache_filename and tuner_config global parameters.
+    #
+    def test_conv_backward_fused(self):
+        conv = """
+        def convolution(float(N,C,H,W) I, float(M,C,KH,KW) W1) -> (O) {
+            O(n, m, h, w) +=!
+                I(n, r_c, h + r_kh, w + r_kw) * W1(m, r_c, r_kh, r_kw)
+        }
+        def convolution_grad(
+            float(N,C,H,W) I, float(M,C,KH,KW) W1, float(N,M,H,W) d_O)
+            -> (d_I, d_W1)
+        {
+            d_I(n, c, h, w) +=!
+                d_O(  n, r_m, h - r_kh, w - r_kw) * W1(r_m, c, r_kh, r_kw)
+            d_W1(m, c, kh, kw) +=!
+                d_O(r_n,   m, r_h - kh, r_w - kw) *  I(r_n, c,  r_h,  r_w)
+        }
+        """
 
-if not os.path.exists(PATH_PREFIX):
-    os.makedirs(PATH_PREFIX)
+        N, C, H, W, O, kH, kW = 32, 4, 56, 56, 16, 1, 1
+        cache_filename = "/tmp/" + str(uuid.uuid4())
```

Comment:
It's generally better to do
```python
with tmpfile.NamedTemporaryFile() as f:
    ...
```
because it guarantees that it will get cleaned up even when an error is raised

Reply:
yay RAII, thanks!